### ISSUE TITLE
go 1.23

### ISSUE THREAD

**Issue by 5k3105** [mappu/miqt#25]
Three things Id like to ask

```
docker run --rm -v $(pwd):/src -w /src miqt/win64-cross:latest go build -buildvcs=false -ldflags '-s -w -H windowsgui'
go: errors parsing go.mod:
/src/go.mod:3: invalid go version '1.23.0': must match format 1.23
```

It did compile though after I edited the go.mod file to say 1.23 instead of 1.23.0

Other people have reported this problem. it isnt a miqt specific problem.

After it compiled, I ran it in wine and got:

![wine_qt_error](https://github.com/user-attachments/assets/87fce1ab-ebeb-4ee2-b9a1-1876eef5d309)

I never had this issue in thereciepe/qt docker builds though.

Then your instructions about GOCACHE:

```
For repeated builds, the compile speed can be improved if you also bind-mount the Docker container's GOCACHE directory: -v ./container-build-cache:/root/.cache/go-build
```

After I compiled it the first time in docker, I ran again like this:

docker run --rm -v $(pwd):/src -w /src miqt/win64-cross:latest go build -buildvcs=false -ldflags '-s -w -H windowsgui' -v ./container-build-cache:/root/.cache/go-build

I got:
stat /src/container-build-cache:/root/.cache/go-build: directory not found

**Comment by 5k3105** [comment]
Constants are missing, for example wont work (from therecipe/qt/ constants) :

```go
window.SetTabPosition(qt.Qt__LeftDockWidgetArea, qt.QTabWidget__North)
```


```go
func (this *QMainWindow) SetTabPosition(areas int, tabPosition uintptr) {
	C.QMainWindow_SetTabPosition(this.h, (C.int)(areas), (C.uintptr_t)(tabPosition))
}
```

Also, I have a QToolBar t:

```go
	t.AddSeparator()
	scan_label = qt.NewQLabel()
	t.AddWidget(scan_label)
```

> cannot use scan_label (variable of type *qt.QLabel) as *qt.QWidget value in argument to t.AddWidget

Also, I like how you shortened ConnectTriggered(func() bool) to OnTriggered1(func()). Thank you

**Comment by mappu** [comment]
Thank you for trying the binding and the very detailed feedback! I can fix all of these issues.

> Constants are missing

I'll solve this soon. See #18 and everything with [tag essential](https://github.com/mappu/miqt/issues?q=is%3Aissue+is%3Aopen+label%3Aessential).

> cannot use scan_label (variable of type *qt.QLabel) as *qt.QWidget value

Qt class inherited types are projected as a Go embedded struct. Therefore, you can write `scan_label.QWidget` to get the QWidget* pointer type: `t.AddWidget(scan_label.QWidget)`.

I'm not sure if there's a more ergonomic way to make this work. Until then, i'll add a note in the README.

**Comment by mappu** [comment]
> invalid go version '1.23.0': must match format 1.23

It's caused because the docker container had go1.19, which only understands the old two-part format. I've added more docker containers with go1.23 that can avoid this issue.

The README instruction has also been updated. You can rebuild the docker container to avoid this issue.

> I ran it in wine and got [...] "no Qt platform plugin could be initialized"

It's now fixed in master with commits cef75b87bdbd307810a77f45e1f3b9152f2c5ac2 and cef75b87bdbd307810a77f45e1f3b9152f2c5ac2. I'll tag a new release soon.

> After I compiled it the first time in docker, I ran again like this:
>
> docker run --rm -v $(pwd):/src -w /src miqt/win64-cross:latest go build -buildvcs=false -ldflags '-s -w -H windowsgui' -v ./container-build-cache:/root/.cache/go-build

Sorry, the `-v` parameter is a docker argument, not a `go build` argument. It should go earlier in the command, such as, before or after the `-v $(pwd):/src` parameter.

I've also updated the example command to use an absolute path instead of a relative path that it seems docker doesn't like.

Try this:

`docker run --rm -v $(pwd)/container-build-cache:/root/.cache/go-build -v $(pwd):/src -w /src miqt/win64-cross:latest go build -buildvcs=false -ldflags '-s -w -H windowsgui'`

After you run with this argument, the cache will be populated. Then, running the command again, will build quickly from the cache.

**Comment by 5k3105** [comment]
Wow. Thank you for responding so quickly! I am very impressed you've taken on this complex project. 
I filed my first issue with therecipe back in 2015 https://github.com/therecipe/qt/issues/4 and used it ever since (back when compile took 17 minutes each time).
So happy you are reviving it




> I'll solve this soon. See #18 and everything with tag essential.

Thank you

> Qt class inherited types are projected as a Go embedded struct. Therefore, you can write scan_label.QWidget to get the QWidget* pointer type: t.AddWidget(scan_label.QWidget).
> I'm not sure if there's a more ergonomic way to make this work. Until then, i'll add a note in the README.


> cannot use model (variable of type *qt.QStandardItemModel) as *qt.QAbstractItemModel value in argument to view.SetModel

This also works if I add .QAbstractItemModel, but therecipe had found a way to handle this automatically.


> It's caused because the docker container had go1.19, which only understands the old two-part format. I've added more docker containers with go1.23 that can avoid this issue.

Thank you for fixing this


> docker run --rm -v $(pwd)/container-build-cache:/root/.cache/go-build -v $(pwd):/src -w /src miqt/win64-cross:latest go build -buildvcs=false -ldflags '-s -w -H windowsgui'

this worked great - thanks!

**Comment by mappu** [comment]
Enum constants are now available in the v0.3.0 tag. 

> have you considered https://github.com/xlab/c-for-go for binding generation?

Before starting this project I considered this, and also many other possible alternatives. The idea to use Clang CLI as the command-line driver was inspired by @jerous86's work on [Nim language bindings](https://github.com/jerous86/nimqt).

My understanding of c-for-go (cgogen) is: it can be used for C++ projects only if they have an extern C interface. Qt doesn't have such C interface officially, so at least half of miqt's work would still be required. Although, there are other Qt C bindings such as [libqt5pas](https://github.com/davidbannon/libqt5pas) `cbindings`, or [kitech/qt.inline](https://github.com/kitech/qt.inline), that might be usable with c-for-go (cgogen).

**Comment by 5k3105** [comment]
> Enum constants are now available in the v0.3.0 tag.

Awesome! I will test it out

> so at least half of miqt's work would still be required

I write all my ui in Qt, but I know there was an attempt at separating the blender ui from the main code base.
The ui has python bindings only. There is also https://github.com/UPBGE/upbge which seems to stay in sync with blender, but may have more features - and can be used to generate standalone apps using the same blender ui - which is gnu gpl.

**Comment by 5k3105** [comment]
> > Enum constants are now available in the v0.3.0 tag.
> 
> Awesome! I will test it out

In v0.3.0, this is the generated binding. 


```go
func (this *QMainWindow) SetTabPosition(areas int, tabPosition QTabWidget__TabPosition) {
	C.QMainWindow_SetTabPosition(this.h, (C.int)(areas), (C.uintptr_t)(tabPosition))
}
```

This is a the code I wrote:

```go
/// therecipe-qt (works)
window.SetTabPosition(qt.QtLeftDockWidgetArea, qt.QTabWidget__North)

/// miqt
window.SetTabPosition(qt.LeftDockWidgetArea, qt.QTabWidget__TabPosition__North)

```

I guess `qt.LeftDockWidgetArea` should work but I cant find the constant and the binding accepts `areas int`.

Docs link: https://doc.qt.io/qt-5/qmainwindow.html#setTabPosition


Thanks you for your work so far.

**Comment by mappu** [comment]
> I guess qt.LeftDockWidgetArea should work but I cant find the constant and the binding accepts areas int.

In miqt it's currently [`qt.DockWidgetArea__LeftDockWidgetArea`](https://github.com/mappu/miqt/blob/master/qt/gen_qnamespace.go#L1089).

I'm interested to see miqt has longer names because therecipe removed the extra name part. Without this part, I think name collisions are possible. If there are no name collisions, then maybe miqt should change to be more compatible.

In most cases for enums, the binding will show the wanted type. But `Qt::DockWidgetAreas` is a `QFlags<enum Qt::DockWidgetArea>`, and I left all `QFlags<T>` types as plain `int` instead of the enum name, because the expectation is that a user ORs multiple flags together. Because this breaks out of the enum choice space, I think both `int` and `qt.DockWidgetArea` are imperfect choices. But probably `qt.DockWidgetArea` is still better, so I'll change this soon.

**Comment by 5k3105** [comment]
sorry, therecipe-qt line was:
```go
window.SetTabPosition(core.Qt__LeftDockWidgetArea, widgets.QTabWidget__North)
```

**Comment by 5k3105** [comment]
using
```go
window.SetTabPosition(qt.DockWidgetArea__LeftDockWidgetArea, qt.QTabWidget__TabPosition__North)
```
i get
`./main.go:52:24: cannot use qt.DockWidgetArea__LeftDockWidgetArea (constant 1 of type qt.DockWidgetArea) as int value in argument to window.SetTabPosition`

wrapping with int fixes it
```go
window.SetTabPosition(int(qt.DockWidgetArea__LeftDockWidgetArea), qt.QTabWidget__TabPosition__North)
```

**Comment by mappu** [comment]
Hi @5k3105 - I've worked on this, and in master and in the next release after v0.4.0, the line should be:

```golang
/// therecipe-qt
window.SetTabPosition(core.Qt__LeftDockWidgetArea, widgets.QTabWidget__North)

// miqt 0.3.0
window.SetTabPosition(int(qt.DockWidgetArea__LeftDockWidgetArea), qt.QTabWidget__TabPosition__North)

// miqt 0.5.0 or master
window.SetTabPosition(qt.LeftDockWidgetArea, qt.QTabWidget__North)
```

The int cast is no longer needed, and the type names are short like therecipe-qt had.

> I'm interested to see miqt has longer names because therecipe removed the extra name part. Without this part, I think name collisions are possible. If there are no name collisions, then maybe miqt should change to be more compatible.

There were name collisions (e.g. in QColorSpace and some others). therecipe had avoided this by skipping the QColorSpace class, it's not present in his binding.

**Comment by mappu** [comment]
The v0.5.0 release I think resolved all points raised in this issue. Thank you for raising everything, MIQT is much better off for it.
